Improve error reporting for union member types

Cito opened the pull request:
First make sure that the included type is an object type, because that would be the more severe problem. Only then we can be sure that the type has a name and check for duplicates. Also, create only one error per different non-object type.

Cito committed:
Improve error reporting for union member types

First make sure that the included type is an object type,
because that would be the more severe problem. Only then we
can be sure that the type has a name and check for duplicates.
Also, create only one error per different non-object type.

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -438,7 +438,20 @@ function validateUnionMembers(
   }
 
   const includedTypeNames = Object.create(null);
+  const includedInvalidTypes = Object.create(null);
```
I think it strange behavior where after fixing one error you would get the same error in the next location.
We should definitely prevent implicit cascading errors but I think we should never mask explicit errors made by the user.
Also if we add this error reduction trick we need to add it in many other places (e.g. if you implement non-interface types) for consistency.

So I think moving check to beginning of the loop and adding `continue` would be enough.

Cito replied:
Note that this does not hide any errors - there is only one message, but all locations are reported and the same message applies to them all.

IvanGoncharov committed:
Fully switch to ESM (scripts, tests, etc.)

yaacovCR committed:
refactor: remove null from executeOperation return type

not actually possible!

IvanGoncharov committed:
ESLint: error on function accepting more than 5 parameters (#3641)

IvanGoncharov committed:
Remove 'devAssert' checks that duplicate TS types (#3642)

`graphql` provides TS types since `14.5.0` (released 3 years ago)
and we fully switched to TS in `15.0.0` so I think it's time to drop
runtime typechecks.

Motivations: This type checks were added long time ago since we shifted
towards TS we just maintained them without adding new ones.
In general, this check increase bundle size add runtime cost and we
can't realistically check all arguments to all functions.
Instead we should focus on adding more asserts on stuff that can't be
checked by TS.

IvanGoncharov committed:
Remove 'assertValidExecutionArguments' function (#3643)

After we drop runtime typechecks that were duplicating TS types this
function became a wrapper for 'assertValidSchema' so all implementations
that rely on it can now use 'assertValidSchema' directly.

yaacovCR committed:
refactor: executeOperation can take only one parameter

`operation` is included within `exeContext`

yaacovCR committed:
refactor: use spread syntax with exeContext

yaacovCR committed:
refactor: subscribe: introduce `buildPerEventExecutionContext` (#3639)

= introduces `buildPerEventExecutionContext` that creates an `ExecutionContext` for each subscribe event from the original `ExecutionContext` used to create the event stream
= `subscribe` now directly builds the `ExecutionContext` instead of relying on `createSourceEventStream`
= introduces `createSourceEventStreamImpl` and `executeImpl` functions that operate on the built `ExecutionContext` rather the `ExecutionArgs`
= `subscribe` calls the `createSourceEventStreamImpl` function on the original context and eventuallys calls `executeImpl` on the per event context created by `buildEventExecutionContext`.

Motivation:
1. remove unnecessary `buildExecutionContext` call, reducing duplicate work
2. paves the way for easily enhancing the `buildPerEventExecutionContext` to add a new `perEventContextFactory` argument to augment the context argument passed to resolvers as need per event.

robrichard committed:
Async benchmarks (#3134)

* support async benchmark tests

* Add benchmarks for sync and async list fields

IvanGoncharov committed:
Support returning async iterables from resolver functions (#2757)

Co-authored-by: Rob Richard <rob@1stdibs.com>

twof committed:
[RFC] Client Controlled Nullability experiment implementation w/o execution (#3418)

Co-authored-by: Ivan Goncharov <ivan.goncharov.ua@gmail.com>

IvanGoncharov committed:
Update deps (#3667)

IvanGoncharov committed:
Cleanup words/files ignores in `cspell.yml` (#3668)

IvanGoncharov committed:
tsconfig: Explicitly list all type checking options

As part of #3649 research I added all options explicitly and also
enable ones that didn't require any code change.
We should enable rest of them in a separate PRs.

IvanGoncharov committed:
tsconfig: Add support for comments

IvanGoncharov committed:
tsconfig: enable `noImplicitOverride` check

IvanGoncharov committed:
Fix: make public API compatible with 'exactOptionalPropertyTypes' (#3670)

IvanGoncharov committed:
Parser: allow 'options' to explicitly accept undefined (#3678)

Continuation of #3670
Basically allow to have code like this:
```js
parse(document, GraphQLServerConfig.parserOptions)
```
to pass TS check with 'exactOptionalPropertyTypes'

spawnia committed:
Fix coerceInputValue test name (#3681)

* Fix coerceInputValue test name

* fix grammar

IvanGoncharov committed:
validate: remove warning about throwing literal (#3680)

yaacovCR committed:
polish: remove unnecessary Maybe (#3683)

IvanGoncharov committed:
parser: limit maximum number of tokens (#3684)

* parser: limit maximum number of tokens

Motivation: Parser CPU and memory usage is linear to the number of tokens in a
document however in extreme cases it becomes quadratic due to memory exhaustion.
On my mashine it happens on queries with 2k tokens.
For example:
```
{ a a <repeat 2k times> a }
```
It takes 741ms on my machine.
But if we create document of the same size but smaller number of
tokens it would be a lot faster.
Example:
```
{ a(arg: "a <repeat 2k times> a" }
```
Now it takes only 17ms to process, which is 43 time faster.

That mean if we limit document size we should make this limit small
since it take only two bytes to create a token, e.g. ` a`.
But that will hart legit documents that have long tokens in them
(comments, describtions, strings, long names, etc.).

That's why this PR adds a mechanism to limit number of token in
parsed document.
Also exact same mechanism implemented in graphql-java, see:
https://github.com/graphql-java/graphql-java/pull/2549

I also tried alternative approach of counting nodes and it gives
slightly better approximation of how many resources would be consumed.
However comparing to the tokens, AST nodes is implementation detail of graphql-js
so it's imposible to replicate in other implementation (e.g. to count
this number on a client).

* Apply suggestions from code review

Co-authored-by: Yaacov Rydzinski  <yaacovCR@gmail.com>

Co-authored-by: Yaacov Rydzinski  <yaacovCR@gmail.com>

IvanGoncharov committed:
Replace all instances of `ObjMap<boolean>` with ES6 Set (#3688)

Motivation: I was inspecting code for unrelated reason and just spotted this pattern.
I think it worth to refactor it since it easier to understand intention looking on code that uses Set's has/add methods.
Benchmark shows no measurable changes in CPU/memory usages.

IvanGoncharov committed:
Remove superficial `.values()` and `.entries()` calls (#3689)

Motivation: Spotted during investigation of unrelated issue.
Maps and sets are iteratable, so no need to call these methods.

IvanGoncharov committed:
Motivation: looked into this it as part of #3687 (#3690)

Explanation: `Object.create(null)` returns value of `any` type.
So bellow construct is not reported by TS even in "strict" mode:
```ts
const foo = Object.create(null);
```
Fixing this issue in `extendSchema` requires adding more code since we
can't put all extensions nodes into one collection without loosing
typesafety.

IvanGoncharov committed:
gen-changelog: Use `fetch` instead of `http.request` (#3692)

IvanGoncharov committed:
Update deps (#3693)

IvanGoncharov committed:
Add workaround to run docusaurus in ESM project

yaacovCR committed:
polish: add tests for `expectPromise` and `expectEqualPromisesOrValues` (#3655)

* polish: add tests for more testUtils

adds tests for expectPromise and expectEqualPromisesOrValues

* break out expectMatchingValues into another file

* apply review feedback

yaacovCR committed:
polish: rename `mapAsyncIterator` to `mapAsyncIterable`

...as the function maps iterables (by extracting the iterator).

yaacovCR committed:
mapAsyncIterable: use `expectPromise` in tests

IvanGoncharov committed:
benchmark: use 'execOutput' from 'util.ts' (#3695)

IvanGoncharov committed:
resources: Move `mkdirSync` into `writeGeneratedFile` (#3697)

IvanGoncharov committed:
resources: merge execOutput and exec into single function (#3698)

IvanGoncharov committed:
ci: add dependency review workflow (#3699)

IvanGoncharov committed:
Workaround for codesandbox having bug with TS enums (#3686)

IvanGoncharov committed:
resources: replace execSync with spawnSync (#3704)

IvanGoncharov committed:
ci: Add CodeQL analysis (#3700)

robrichard committed:
Reference implementation of defer and stream spec (#3659)

Co-authored-by: Liliana Matos <liliana@1stdibs.com>
Co-authored-by: David Glasser <glasser@davidglasser.net>

IvanGoncharov committed:
ci: Move 'integration-test.ts' into 'resources' (#3721)

IvanGoncharov committed:
resources: add utility function to create tmp dir (#3723)

IvanGoncharov committed:
resources/utils.ts: npm/git will inherit stdout by default (#3724)

robrichard committed:
Filter subsequent payloads when parent field is null (#3720)

yaacovCR committed:
update incremental docs (#3731)

yaacovCR committed:
incrementalDelivery: remove singleResult wrapper (#3733)

extracted from #3732

IvanGoncharov committed:
changelog: use top-level await to simplify code (#3725)

IvanGoncharov committed:
resources: replace usage of deprecated ts function (#3735)

IvanGoncharov committed:
promiseForObject: convert to be async function (#3736)

IvanGoncharov committed:
Update deps (#3737)

IvanGoncharov committed:
Remove unused Babel config files (#3738)

yaacovCR committed:
incrementalDelivery: fix iterable streaming with errors (#3729)

IvanGoncharov committed:
build-npm/deno: use TS to find all *.ts files (#3741)

robrichard committed:
fix github actions lint (#3744)

PabloSzx committed:
deploy full ESM with "type": "module" in "latest-esm" npm tag (#3361)

IvanGoncharov committed:
resources/util: spawn command fail on error (#3745)

IvanGoncharov committed:
17.0.0-alpha.2

yaacovCR committed:
add filtering for async iterator lists

yaacovCR committed:
fix: allow async errors to bubble to AsyncIterable list items (#3746)

sashashura committed:
GitHub Workflows security hardening (#3751)

yaacovCR committed:
polish: add additional test for filtering (#3749)

Filtering is still required for synchronous error bubbling with stream.

With defer, patches are not executed if the initial field set execution fails, but with stream, the initial items (perhaps of zero length) will execute and will create the stream record.

Cito committed:
Fix typo in the error message for max tokens (#3748)

yaacovCR committed:
incrementalDelivery: refactoring and streamlining (#3728)

yaacovCR committed:
fix(incrementalDelivery): filtering should never filter the error source

use explicit undefined to ensure that all code points are covered

yaacovCR committed:
polish(incrementalDelivery): filter function is always passed a path

yaacovCR committed:
refactor: introduce completeListItemValue (#3752)

yaacovCR committed:
fix: re-initialize subsequentPayloads (#3757)

chrskrchr committed:
Fix crash in node when mixing sync/async resolvers (#3706)

Co-authored-by: Ivan Goncharov <ivan.goncharov.ua@gmail.com>

spawnia committed:
Fix formatting in schema for test "resolve Interface type using __typename on source object" (#3762)

IvanGoncharov committed:
resources/utils: fix `git().revList` (#3763)

yaacovCR committed:
fix(github-actions-bot): add additional permissions (#3759)

IvanGoncharov committed:
Update deps (#3766)

yaacovCR committed:
fix(incrementalDelivery): fix null bubbling with async iterables (#3760)

igrlk committed:
Fix typo in CONTRIBUTING.md (#3769)

Cito committed:
Minor simplification in schema-test (#3770)

IvanGoncharov committed:
lint: enable ESLint's `logical-assignment-operators` rule (#3771)

IvanGoncharov committed:
eslintrc: reorder rules to match order in the official docs (#3772)

IvanGoncharov committed:
eslintrc: enable few rules with minimal code changes (#3774)

robrichard committed:
Remove outdated instructions on experimental tag (#3778)

robrichard committed:
Fix comments in defer/stream validation rules (#3781)

IvanGoncharov committed:
Update deps (#3779)

igrlk committed:
extendSchema: preserve "description" and "extensions" (#3776)

Co-authored-by: Ivan Goncharov <ivan.goncharov.ua@gmail.com>
Closes https://github.com/graphql/graphql-js/issues/3629

IvanGoncharov committed:
workflows: fix and cleanup permissions (#3797)

yaacovCR committed:
fix permissions for codeql-action (#3800)

yaacovCR committed:
polish: do not repeat isPromise check (#3793)

yaacovCR committed:
enhancement: remove extra ticks (#3754)

Promise resolution order changes in some instances, resulting in different orders for some errors within the errors array, as well as in different values of hasNext within incremental delivery payloads.

This PR introduces an async `completePromisedValue` helper function rather than using a promise chain (see below links).

https://github.com/tc39/proposal-faster-promise-adoption
https://github.com/tc39/ecma262/issues/2770
https://github.com/tc39/ecma262/pull/2772
https://github.com/tc39/ecma262/pull/1250
https://v8.dev/blog/fast-async

Depends on #3793

yaacovCR committed:
polish: narrow mapSourceToResponse return type (#3801)

robrichard committed:
Remove defer/stream support from subscriptions (#3742)

yaacovCR committed:
original `execute` should throw if defer/stream directives are present (#3722)

yaacovCR committed:
polish: remove misleading comment (#3806)

We could not demonstrate via actual benchmarking that the tick reduction
from awaiting a promise prior to returning it translates into real-world
performance gains. See #3796.

In any case, the promise must be awaited (irrespective of performance)
to allow the catch block to handle rejection errors. Simply returning
`completed` would result in test failures.

yaacovCR committed:
polish: clarify filtering test semantics (#3816)

`expect.fail()` was mistakenly introduced in #3760 under the assumption that expect.fail() will always cause the test to fail, and that `.next()` should be called at most once.

Actually, thought, `expect.fail()` just throws an error, `.next()` is called more than once, the error is produced and wrapped by GraphQL, but then it is filtered, so the test ultimately passes.

Fortunately, that's actually the desired behavior! It's ok if the number of calls to `.next()` is more than 1 (in our case it is 2). The exact number of calls to `.next()` depends on the # of ticks that it takes for the erroring deferred fragment to resolve, which should be as low as possible, but that is a separate objective (with other PRs in the mix already aimed at reducing).

So the test as written is intending to be overly strict, but is not actually meeting that goal and so functions as desires. This PR makes no change to the test semantics, but changes the comment, the coverage ignore statement, and removes the potentially confusing use of `expect.fail`, so that the test semantics are more clearly apparent.

spawnia committed:
Avoid an before consonant (#3823)

Cito committed:
Fix wording of comment in buildClientSchema (#3821)

stenreijers committed:
fix: invalid originalError propagation in custom scalars (#3837)

yaacovCR committed:
internal: fix website build error (#3843)

internal eslint rules are referenced in package json and required to be copied for a proper install.

to ponder: why did this ever work?

benjie committed:
Fix ambiguity around when schema definition may be omitted (#3839)

This PR implements the tests and fixes necessary for [Spec RFC
#987](https://github.com/graphql/graphql-spec/pull/987).

This PR is made of three main commits:

1. Test printing a schema that has `Query`, `Mutation` and `Virus`
types, but only supports `query` operations (via the `Query` type) and
does _not_ support `mutation` operations.
2. Test parsing this same schema text, and assert that the schema does
not have a mutation type.
3. Fix the printing of the schema.

Co-authored-by: Lee Byron <lee@leebyron.com>

IvanGoncharov committed:
ESLint: use the `--rulesdir` option for internal rules. (#3849)

yaacovCR committed:
internal: fix inlineInvariant for cjs builds that require filenames with explicit extensions (#3832)

IvanGoncharov committed:
Add node@19 to integration tests + use it in CI (#3848)

Motivation: to run it as part of integration tests and also use it in CI

IvanGoncharov committed:
ci: uncomment package-lock.json check for unmet dependencies (#3850)

IvanGoncharov committed:
ESLint: switch to '.eslintrc.cjs' as config file (#3852)

IvanGoncharov committed:
Update deps (#3854)

IvanGoncharov committed:
Drop support for node14 (#3855)

IvanGoncharov committed:
Use `replaceAll` instead of RegExp with global flag (#3856)

Cito committed:
Add accidentally deleted test spec name back (#3865)

dylanowen committed:
Narrow the return type of astFromValue (#3864)

IvanGoncharov committed:
Use ES2021 `Array.at` (#3857)

IvanGoncharov committed:
printSchema: correctly print empty description (#3869)

IvanGoncharov committed:
Enable @typescript-eslint/strict-boolean-expressions rule (#3872)

Motivation: Fix edge cases like in #3869
Also, I noticed a similar issue in #3867, so I fixed it for the entire codebase.

P.S. I also resulted in a small but measurable speedup, probably because I replaced a bunch of ObjMap with ES6 Map.

IvanGoncharov committed:
ASTReducer: correctly type scalar values (#3873)

IvanGoncharov committed:
GraphQLObjectType: partial revert of #3143, due to perf regression (#3874)

IvanGoncharov committed:
GraphQLInputObjectType: remove check that duplicate TS types (#3875)

yaacovCR committed:
introduce FieldGroup type

yaacovCR committed:
rename fieldNodes variable to fieldGroup

to match type

yaacovCR committed:
introduce GroupedFieldSet type

yaacovCR committed:
use groupedFieldSet as variable name

to match type

yaacovCR committed:
refactor handleFieldError

= integrate locatedError, which was always called on the input
= avoid constant return value, which causes return value to hide value of this constant
= add optional asyncPayloadRecord parameter so that caller does not have to supply errors list

IvanGoncharov committed:
GraphQL*Type: bind only single property instead of entire context (#3876)

yaacovCR committed:
rename executeStreamIterator

to executeStreamAsyncIterator

IvanGoncharov committed:
GraphQLInterface: add missing template parameters (#3877)

dylanowen committed:
Make print() break long List and Object Values over multiple line (#3860)

Co-authored-by: Ivan Goncharov <ivan.goncharov.ua@gmail.com>

IvanGoncharov committed:
gen-changelog: Fix error introduced in #3872 (#3879)

robrichard committed:
simplify schema in defer tests (#3881)

Cito committed:
Testing the right functions in testUtils (#3880)

tomgasson committed:
Expose printDirective function to enable schema sharding (#3822)

IvanGoncharov committed:
eslint: replace unmaintained 'node' plugin with it's fork (#3882)

yaacovCR committed:
remove unnecessary duplicated fields from defer tests (#3885)

extracted from #3862

yaacovCR committed:
remove some optionality to parameters

yaacovCR committed:
rename fields parameter in executeFieldsSerially to groupedFieldSet

to match type

Cito committed:
Fix misleading test section description (#3890)

yaacovCR committed:
rename StreamRecord to StreamItemsRecord

yaacovCR committed:
rename iterator to asyncIterator

yaacovCR committed:
rename AsyncPayloadRecord to IncrementalDataRecord

IvanGoncharov committed:
update deps (#3884)

yaacovCR committed:
executeFields: update grouped field set variable name (#3896)

missed by me both in #3891 and #3878

yaacovCR committed:
asyncIterable: locate async iterator errors to the collection (#3899)

Currently, they are presented as errors in the next pending item.

Of course, there may or not be a next pending item; there may not be a
way to definitively know the reason for the error, such that the error
is a function of a problem with the collection itself.

As discussed in a prior incremental delivery WG meeting, this must be
changed for both the non-streaming and streaming cases; this PR makes
the minor adjustments addressing both cases.

robrichard committed:
introduce promiseWithResolvers helper (#3902)

Promise.withResolvers is currently a [stage 2 proposal for
JS](https://github.com/tc39/proposal-promise-with-resolvers). This adds
an equivalent jsutil to standardize usage.

yaacovCR committed:
execute: move publishing code into separate file (#3903)

yaacovCR committed:
introduce new IncrementalPublisher class (#3894)

extracted from #3886

depends on #3903

more refactors from the without-branching branch, a bit more fundamental
than #3891

[set as patch release because it does have an observable effect on the
number of payloads, see below]

= iterate only through completed items
= remove extra ticks by making the publisher manage changes to its state
synchronously.
= use children array instead of promises to manage hierarchy
= have IncrementalPublisher instantiate new IncrementalDataRecords

= The new publisher sometimes cause an empty `{ hasNext: false }` to be
emitted. In particular, because the publisher is faster than it was, it
may emit a stream result before the stream's asynchronous iterator has
completed.
= The new publisher may sometimes reduce the number of `{ hasNext: false
}` records that are emitted. For example, when errors on the initial
result filter all subsequent results, this now happens synchronously,
and so the publisher knows immediately that there are no subsequent
results, such that there is no need for an empty final payload.

erikkessler1 committed:
Implement OneOf Input Objects via `@oneOf` directive (#3513)

Co-authored-by: Benjie Gillam <benjie@jemjie.com>
Closes graphql/graphql-wg#648

yaacovCR committed:
incremental: subsequent result records should not store parent references (#3929)

as memory then cannot be freed

This affects both the existing branching executor on main as well as the
non-branching, deduplicated version in #3886

We want to ensure that after an incremental result is sent to the
client, no subsequent results reference this result so that garbage
collection can free the memory associated with the result. To effect
this, two changes are required:

1. Prior to this change, we performed filtering by modifying the
children stored on a parent; now we add a flag on the filtered item
itself, so that we no longer need a backreference from child to parent.
2. We no longer store a permanent reference on the ExecutionContext to
the children of the initial result. Rather, we have the
IncrementalPublisher provide a new InitialResultRecord that carries its
own errors and children properties.

IvanGoncharov committed:
Speedup sorting & building/extending schema (#3939)

benjie committed:
Fix edit URL in documentation (#3947)

gschulze committed:
Add support for fourfold nested lists in introspection (#3948)

Closes #2643

sakesun committed:
Fix invalid sample code (#3951)

yaacovCR committed:
incremental publisher should handle all response building (#3930)

extracted from #3929

the publisher itself can determine whether to return a single result or
the initial result + stream

the only desired change is to replace the following code block with the
below:


[FROM:](https://github.com/graphql/graphql-js/blob/fae5da500bad94c39a7ecd77a4c4361b58d6d2da/src/execution/execute.ts#L293-L340)

```ts
  const incrementalPublisher = exeContext.incrementalPublisher;
  const initialResultRecord = incrementalPublisher.prepareInitialResultRecord();
  try {
    const result = executeOperation(exeContext, initialResultRecord);
    if (isPromise(result)) {
      return result.then(
        (data) => {
          const errors =
            incrementalPublisher.getInitialErrors(initialResultRecord);
          const initialResult = buildResponse(data, errors);
          incrementalPublisher.publishInitial(initialResultRecord);
          if (incrementalPublisher.hasNext()) {
            return {
              initialResult: {
                ...initialResult,
                hasNext: true,
              },
              subsequentResults: incrementalPublisher.subscribe(),
            };
          }
          return initialResult;
        },
        (error) => {
          incrementalPublisher.addFieldError(initialResultRecord, error);
          const errors =
            incrementalPublisher.getInitialErrors(initialResultRecord);
          return buildResponse(null, errors);
        },
      );
    }
    const initialResult = buildResponse(result, initialResultRecord.errors);
    incrementalPublisher.publishInitial(initialResultRecord);
    if (incrementalPublisher.hasNext()) {
      return {
        initialResult: {
          ...initialResult,
          hasNext: true,
        },
        subsequentResults: incrementalPublisher.subscribe(),
      };
    }
    return initialResult;
  } catch (error) {
    incrementalPublisher.addFieldError(initialResultRecord, error);
    const errors = incrementalPublisher.getInitialErrors(initialResultRecord);
    return buildResponse(null, errors);
  }
}
```


[TO:](https://github.com/yaacovCR/graphql-executor/blob/598608e8d8b23bc527dd73283b477997305afd58/src/execution/execute.ts#L234-L250):

```ts
  const incrementalPublisher = exeContext.incrementalPublisher;
  const initialResultRecord = incrementalPublisher.prepareInitialResultRecord();
  try {
    const data = executeOperation(exeContext, initialResultRecord);
    if (isPromise(data)) {
      return data.then(
        (resolved) =>
          incrementalPublisher.buildDataResponse(initialResultRecord, resolved),
        (error) =>
          incrementalPublisher.buildErrorResponse(initialResultRecord, error),
      );
    }
    return incrementalPublisher.buildDataResponse(initialResultRecord, data);
  } catch (error) {
    return incrementalPublisher.buildErrorResponse(initialResultRecord, error);
  }
```

Supporting changes are required:
1. some existing public methods no longer are required to be public and
so are made private (or removed entirely!), with lint rules forcing the
reordering of existing methods
2. to prevent cyclic type dependencies (not strictly necessary, but
still!), types are moved from `execute.ts` to `IncrementalPublisher.ts`

yaacovCR committed:
add tests with regard to duplication (#3904)

to minimize later diffs, we can watch how the tests change, rather than
simple introducing tons of new tests

depends on https://github.com/graphql/graphql-js/pull/3930

yaacovCR committed:
Introduces new incremental response format  (#3886)

The `incremental` array is the actual data to be applied to the response, while the `completed` array return "metadata" about the execution, used to inform clients that defers are being executed and when all fields for that defer have been delivered.

To ensure consistency, clients are expected to process all objects in the `incremental` array for a given payload before re-rendering the associated UIs.

See https://github.com/graphql/defer-stream-wg/discussions/69 for full details.

yaacovCR committed:
incremental delivery: add pending notifications (#3897)

yaacovCR committed:
fix(types): path is required within incremental results (#3961)

extracted from #3960

thanks @benjie for pointing this out!

yaacovCR committed:
incremental: utilize `id` and `subPath` rather than `path` and `label` (#3960)

updates response format to fully match new format further discussed at
https://github.com/graphql/defer-stream-wg/discussions/69

yaacovCR committed:
skip unnecessary initialization of empty items array (#3962)

We can use a non-null assertion to access `items ` on the completed
StreamItemsRecord, similar to how `data` is accessed with a non-null
assertion for completed DeferredGroupedFieldSet records, avoiding
initialization of empty arrays/objects simply to avoid type errors.

spawnia committed:
Improve description for `@oneOf` directive (#3937)

sakesun committed:
fix for development on Windows (#3964)

IvanGoncharov committed:
instanceOf: workaround bundler issue with process.env (#3965)

Port of #3923 from v16 to main (v17)

yaacovCR committed:
polish: improve addDeferredFragments readability (#3966)

-- improve variable naming:

`parentDeferUsage` => `parentTarget`
`deferUsage` => `newDeferUsage`

-- add comments

AaronMoat committed:
OverlappingFieldsCanBeMergedRule: Fix performance degradation (#3958)

Co-authored-by: AaronMoat <AaronMoat@users.noreply.github.com>
Co-authored-by: Ivan Goncharov <ivan.goncharov.ua@gmail.com>
Resolves https://github.com/graphql/graphql-js/issues/3955 (at least

colinhacks committed:
docs: add installation instructions for bun (#3957)

IvanGoncharov committed:
17.0.0-alpha.3

IvanGoncharov committed:
tests: remove empty strings in string literals (#3970)

yaacovCR committed:
fix: suppress records for deferred fragments that are completely empty (#3984)

i.e. no fields and no enclosed deferred fragments

These fragments can be thought of to be completely skipped, because
including them will just result in emitting metadata but no actual data.
Alternatively, these fragments can be thought of as being inlined.

This could probably be considered a bug fix, in that Example F @
https://github.com/graphql/defer-stream-wg/discussions/69 explicitly
states that these fragments should be skipped.

yaacovCR committed:
fix(incremental): fix logic around selecting id/subPath (#3987)

current loop assumes that the first analyzed DeferredFragmentRecord has been released as pending and thereofre has an `id`, which happens to be the case in this scenario, but is not reliable, uncovered by #3982.

Demonstrates the peril of `!` which might point to requiring additional types, but we can leave that for a different PR.

Another option is to include an `invariant` call -- that should be unnecessary, but would indeed be safer.

hayes committed:
Add support for @oneOf directives in printSchema (#3969)

fixes #3968

robrichard committed:
ci: Use node v20 for CI (#3992)

yaacovCR committed:
simplify CollectFields for `@defer` and `@stream` (#3994)

minimizes the changes to `CollectFields` required for incremental delivery (inspired by #3982) -- but retains a single memoized incremental field plan per list item.

saihaj committed:
docs: reviewer and merge process (#4013)

In GraphQL.js working group we discussed and put process in place. But
we never officially wrote down in docs. You can watch this recording
https://youtu.be/BBxc_HmB71I?si=cvdUMI47vBsT9F2y&t=1316 or check the
transcript. These are the changes we agreed in the working group, just
putting them in writing.

---------

Co-authored-by: Benjie <benjie@jemjie.com>

yaacovCR committed:
fix(incremental): skip all empty subsequent results (#3993)

The publish method checks to see if a subsequent result is empty; this
same logic should be employed to suppress pending notices for empty
records.

This has already been achieved for subsequent results that are children
of the initial result, as we generated the pending notices from the list
of initially published records.

For subsequent results that are children of other subsequent results, we
previously generated the pending notice prior to actually publishing.

This change integrates the logic: the publishing method itself returns a
pending notice as required. This results in a bug-fix for subsequent
records of other subsequent records as well as a reduction of code for
subsequent results to the initial result.

yaacovCR committed:
incremental: allow nested defers at the same level (#4002)

Implements Option A from
https://github.com/graphql/defer-stream-wg/discussions/80

yaacovCR committed:
fix(benchmark): select npm.cmd on windows (#4030)

see: https://github.com/nodejs/node/issues/3675
and in particular:
https://github.com/nodejs/node/issues/3675#issuecomment-308963807

yaacovCR committed:
chore(benchmark): update v8 flag (#4029)

changed in v8 11.3 / node 20

see:
https://chromium.googlesource.com/v8/v8/+/73c45833472086ca1c6ab4140cdb058d1890e930%5E%21/src/flags/flag-definitions.h

fixes benchmarking, see
https://github.com/graphql/graphql-js/actions/runs/8314729475/job/22752257524#step:6:142

yaacovCR committed:
polish: rename executeImpl to executeOperation (#4045)

inline much of the original executeOperation
leave out executeRootGroupedFieldSet, will help reduce the diff for
#4026

Cito committed:
Fix stream directive validation error message (#4044)

Fixes the error message created by the
DeferStreamDirectiveOnValidOperationsRule for the stream directive.

yaacovCR committed:
incremental: introduce GraphQLWrappedResult to avoid filtering (#4026)

following https://github.com/graphql/graphql-spec/pull/1077

now part of the following PR stack, with the laters PRs extracted from
this one

#4026: incremental: introduce GraphQLWrappedResult to avoid filtering
#4050: perf: allow skipping of field plan generation
#4051: perf: introduce completePromisedListItemValue
#4052: refactor: introduce completeIterableValue
#4053: perf: optimize completion loops
#4046: perf: use undefined for empty

yaacovCR committed:
perf: allow skipping of field plan generation (#4050)

in the non-deferred case.

depends on #4026

---------

Co-authored-by: Rob Richard <rob@1stdibs.com>

yaacovCR committed:
perf: introduce completePromisedListItemValue  (#4051)

depends on #4050

yaacovCR committed:
refactor: introduce completeIterableValue (#4052)

refactoring that will streamline when we introduce two versions of this
function to optimize the loop when not streaming

depends on #4051

yaacovCR committed:
incremental: avoid double loop with stream from sync iterables (#4076)

yaacovCR committed:
perf: use undefined for empty (#4046)

As convention, we will lazily instantiate arrays/sets when adding the first item.

This applies to arrays/sets on execution/incremental context, as well as the second member of the GraphQLWrappedResult tuple holding the array of incremental data records.

yaacovCR committed:
Add test for consolidating grouped field sets properly into deferred fragments (#3997)

Group field sets should be properly consolidated when some of the fields
in a sibling defer overlap with a child deferred fragment.

The child deferred fragment should not prompt creation of an additional incremental data record for that field, because the field would always be sent with the parent.

JoviDeCroock committed:
fix: remove `globalThis` check and align with what bundlers can accept (#4022)

As surfaced in
[Discord](https://discord.com/channels/625400653321076807/862957336082645006/1206980831915282532)
this currently is a breaking change in the 16.x.x release line which is
preventing folks from upgrading towards a security fix. This PR should
result in a patch release on the 16 release line.

This change was originally introduced to support CFW and browser
environments which should still be supported with the `typeof` check CC
@n1ru4l

This also adds a check whether `.env` is present as in the DOM using
`id="process"` defines that as a global which we don't want to access on
accident. as shown in https://github.com/graphql/graphql-js/pull/4017

Bundles also target `process.env.NODE_ENV` specifically which fails when
it replaces `globalThis.process.env.NODE_ENV` as this becomes
`globalThis."production"` which is invalid syntax.

Fixes https://github.com/graphql/graphql-js/issues/3978
Fixes https://github.com/graphql/graphql-js/issues/3918
Fixes https://github.com/graphql/graphql-js/issues/3928
Fixes https://github.com/graphql/graphql-js/issues/3758
Fixes https://github.com/graphql/graphql-js/issues/3934

This purposefully does not account for
https://github.com/graphql/graphql-js/issues/3925 as we can't address
this without breaking CF/plain browsers so the small byte-size increase
will be expected for bundled browser environments. As a middle ground we
did optimise the performance here. We can revisit this for v17.

Most bundlers will be able to tree-shake this with a little help, in
https://github.com/graphql/graphql-js/issues/4075#issuecomment-2094052098
you can find a conclusion with a repo where we discuss a few.

- Next.JS by default replaces
[`process.env.NODE_ENV`](https://github.com/vercel/next.js/blob/b0ab0fe85fe8c93792051b058e060724ff373cc2/packages/next/webpack.config.js#L182)
you can add `typeof process` linearly
- Vite allows you to specify
[`config.define`](https://vitejs.dev/config/shared-options.html#define)
- ESBuild by default will replace `process.env.NODE_ENV` but does not
support replacing `typeof process`
- Rollup has a plugin for this
https://www.npmjs.com/package/@rollup/plugin-replace

Supersedes https://github.com/graphql/graphql-js/pull/4021
Supersedes https://github.com/graphql/graphql-js/pull/4019
Supersedes https://github.com/graphql/graphql-js/pull/3927

> This now also adds a documentation page on how to remove all of these

yaacovCR committed:
refactor: extract execution types to separate file (#4099)

yaacovCR committed:
refactor: extract incremental graph to separate file

yaacovCR committed:
refactor: move returnStreamIterators to method

yaacovCR committed:
refactor: use asyncIterator instead of extra promise

yaacovCR committed:
refactor: convert IncrementalPublisher class members to method args

yaacovCR committed:
incremental: use invariant for checking id

yaacovCR committed:
refactor(IncrementalGraph): remove unnecessary method

yaacovCR committed:
refactor(IncrementalGraph): use Subsequent Result nodes to reduce mutation

yaacovCR committed:
refactor(IncrementalGraph): use set of pending deferred grouped field set results to reduce mutation

yaacovCR committed:
refactor(incremental): enqueue only released records

yaacovCR committed:
refactor(incremental): introduce BoxedPromiseOrValue to save resolved promise results

yaacovCR committed:
fix(incremental): emit only single completion when multiple deferred grouped field sets error

yaacovCR committed:
refactor(incremental): introduce StreamRecord type

yaacovCR committed:
incremental(stream): revert test logic (#4101)

Test logic was inadvertently altered in #4026.

Tests still pass, just fixing logic.

yaacovCR committed:
polish: remove outdated coverage ignore statement

benjie committed:
Update release instructions in CONTRIBUTING.md (#4105)

yaacovCR committed:
incremental: handle Stream as stream rather than linked list (#4098)

The incremental graph can handle a stream as a stream, rather than
creating a linked list where each incremental data record also includes
the next record in addition to any new defers and/or streams.

Enables easily batching all available stream items within the same
incremental entry.

Depends on #4094

yaacovCR committed:
incremental: disable early execution by default (#4097)

depends on #4098

yaacovCR committed:
polish(IncrementalPublisher): remove unnecessary check and method call (#4106)

polished IncrementalPublisher:

= completeDeferredFragment will always return undefined if the
deferredFragmentRecord has already been removed from the graph
= removeDeferredFragment need not be called twice

benjie committed:
Introduce "recommended" validation rules (#4118)

Co-authored-by: enisdenjo <badurinadenis@gmail.com>
Co-authored-by: enisdenjo <denis@denelop.com>

benjie committed:
port[v17]: Fix publish scripts (#4121)

benjie committed:
17.0.0-alpha.4

benjie committed:
Enable passing values configuration to GraphQLEnumType as a thunk (#4018)

maciesielka committed:
Add `@oneOf` support to introspection query (#4078)

benjie committed:
17.0.0-alpha.5

jbellenger committed:
Require non-empty directive locations (#4100)

benjie committed:
Rename `inputObjectOneOf` to `oneOf` in `getIntrospectionQuery()` (#4125)

yaacovCR committed:
polish(incremental): refactor getNewPending functionality (#4108)

return newPending alongside incremental results when completing a
fragment

= removes need to track `newPending` and `newIncrementalDataRecords` as
part of the class
= unifies handling of retrieving new pending for initial result,
completed fragments, and stream items incremental entries

saihaj committed:
chore: Tag GraphQL JS reviewers team for reviews on PRs (#4130)

yaacovCR committed:
refactor(incremental): simplify incremental graph by allowing mutations (#4112)

depends on #4108

yaacovCR committed:
polish(incremental): replace iterator with current and next accessors

-- we do not need a formal iterator as we are the only consumer.

yaacovCR committed:
polish(incremental): improve IncrementalPublisher return test

Use a stream that never resolves to better demonstrate how there will be a definite pending promise.

Calling return on the underlying graph flushes the nextQueue anyway.

yaacovCR committed:
polish(incremental): move check for empty state to end of batch

Reduces the number of times we need to check as to whether to flush the next queue.

yaacovCR committed:
return async iterables in the non incremental delivery case (#4144)

catching errors

yaacovCR committed:
fix(incremental): do not initiate non-pending execution groups (#4140)

Currently, when early execution is disabled, we still use the early
execution logic to initiate execution groups, which may cause early
initiation of non-pending execution groups.

alternative to #4109, causes potentially stacking delays when
combinations of shared and non-shared keys between sibling defers cause
separate deferred grouped field sets for the same deferred fragment.

yaacovCR committed:
incremental: optimize build field plan (#4147)

main improvement is to remove additional intermediate variables, i.e.
`map`, `inOriginalResult`.

yaacovCR committed:
polish(incremental): rename terms to match spec (#4148)

DeferredGroupedFieldSet => ExecutionGroup
SubsequentResultRecord => DeliveryGroup
ExecuteExecutionGroups => CollectExecutionGroups

yaacovCR committed:
polish(incremental): remove unnecessary helper

yaacovCR committed:
polish(incremental): improve readability with destructuring

yaacovCR committed:
polish(incremental): remove outdated reference to subsequentResult

bignimbus committed:
Add GraphQLConf 2024 banner (#4157)

Co-authored-by: Saihajpreet Singh <saihajpreet.singh@gmail.com>

yaacovCR committed:
polish(incremental): remove unnecessary helper

improves readability at the cost of pushing and then immediate popping

yaacovCR committed:
fix(incremental): fix paths for subsequent async stream items (#4160)

yaacovCR committed:
chore(actions): upgrade to latest versions (#4162)

JoviDeCroock committed:
17.0.0-alpha.7

yaacovCR committed:
internal: remove stryker mutation testing (#4163)

currently not actively being used for code quality improvement — if found useful, can be restored

hayes committed:
fix: improve handling of promises in defaultTypeResolver (#3494)

Currently it is possible to have unhandled promise rejections that arise
from a mix of sync and async isTypeOf checks.


This should fix that issue.

yaacovCR committed:
polish: introduce executeExecutionPlan to conform to spec algo (#4178)

Previously, this was inlined into both executeOperation and
collectAndExecuteSubfields, now it is a separate function.

executeExecutionPlan calls "getNewDeferMap" instead of
"addNewDeferredFragments" to also conform to the spec
functional/immutable style.


![image](https://github.com/user-attachments/assets/95918a1d-e024-411a-87a7-f3835e26a810)

yaacovCR committed:
improve errors around around dots (#4176)

extracted from @leebyron 's excellent #3807

will reduce that diff

yaacovCR committed:
standardize error messages prior to introducing schema coordinates (#4177)

extracted from #3808

PR #3808 uses schema coordinates to improve GraphQL-JS error messages.
To reduce the size of the PR, this commit standardizes error messages
according to the general pattern that will be introduced with schema
coordinates without introducing the coordinates themselves, in the hopes
of aiding review of the later PR.

EDITED 8/26/2024:

I was able to reproduce all of the standardized error messages from
#3808 except for the ones in getArgumentValues when it is passed a Field
Definition, because the parent type is not passed. Everything else can
be calculated for the error messages we are currently printing, although
schema coordinates simplifies things.

Extracting these changes out of #3808 and rebasing #3808 on main will
therefore will better demonstrate how schema coordinates improves the
clarity of some of our error messages (namely, getArgumentValues) and
simplifies printing them.

JoviDeCroock committed:
feat: add experimental support for parsing fragment arguments (#4015)

This is a rebase of https://github.com/graphql/graphql-js/pull/3847

This implements execution of Fragment Arguments, and more specifically
visiting, parsing and printing of fragment-spreads with arguments and
fragment definitions with variables, as described by the spec changes in
https://github.com/graphql/graphql-spec/pull/1081. There are a few
amendments in terms of execution and keying the fragment-spreads, these
are reflected in https://github.com/mjmahone/graphql-spec/pull/3

The purpose is to be able to independently review all the moving parts,
the stacked PR's will contain mentions of open feedback that was present
at the time.

- [execution changes](https://github.com/JoviDeCroock/graphql-js/pull/2)
- [TypeInfo & validation
changes](https://github.com/JoviDeCroock/graphql-js/pull/4)
- [validation changes in
isolation](https://github.com/JoviDeCroock/graphql-js/pull/5)


CC @mjmahone the original author

---------

Co-authored-by: mjmahone <mahoney.mattj@gmail.com>
Co-authored-by: Yaacov Rydzinski <yaacovCR@gmail.com>

yaacovCR committed:
Add flexibility around the waiting period for merging PRs (#4172)

To simply suggest a larger cohort of reviewers for more complex changes, and add the caveat that merged changes may be reverted if consensus has not been reached at a GraphQL-JS WG meeting.

yaacovCR committed:
fix(fragmentArguments): arguments on directives for nested fields (#4180)

fixes up just merged #4015

this was actually intended to be in #4015, but due to branch confusion
not originally included

now we also have a test!

yaacovCR committed:
chore(deps): upgrade dependencies (#4166)

Errata:

Left typescript at v5.4.x as deep dependency typedoc requires v5.4.x.
Left eslint v8.x as not all plugins are compatible with v9.

Code changes, in no particular order:

1. Prettier formatting changes.
2. Prettier moved to an async API, but the `writeGeneratedFile` utility, which previously included prettifying was passed as a callback function to TS and had to stay sync, so prettifying was separated into a separate async function -- the callback function luckily did not seem to actually requiring another round of prettifying, as it just involved renaming. All the other callsites of the new utility had to be made async. In the alternate, I investigated @prettier/sync and the lower-level make-synchronized and make-synchronous packages, but I could not get them working.
3. Plenty of eslint rule changes! I have tried to make sure that the rule list orders now match the linked documentation, so that further updates might be easier.
4. Minor docusaurus config tweaks to get the build to pass.

yaacovCR committed:
chore(ensureValidRuntimeType): remove special v16 error message (#4183)

yaacovCR committed:
remove deprecated custom typeInfo arg for validate() (#4187)

yaacovCR committed:
fix(IntrospectionType): properly type using TypeKind enum (#4185)

yaacovCR committed:
remove deprecated getFieldDefFn arg to TypeInfo (#4186)

yaacovCR committed:
update comments regarding fragment arguments and variables (#4184)

yaacovCR committed:
chore(deps): remove prettier-ignore (#4191)

fixed by deps upgrade

yaacovCR committed:
Add coerceInputLiteral() (#3809)

Deprecates `valueFromAST()` and adds `coerceInputLiteral()` as an additional export from `coerceInputValue`.

The implementation is almost exactly the same as `valueFromAST()` with a slightly more strict type signature .

`coerceInputLiteral()` and only `coerceInputLiteral()` properly supports fragment variables in addition to operation variables.

yaacovCR committed:
fix(validation): catch OverlappingFieldsCanBeMergedRule violations with nested fragments (#4168) (#4199)

yaacovCR committed:
Preserve defaultValue literals (#3810)

[#3074 rebased on
main](https://github.com/graphql/graphql-js/pull/3074).

Depends on #3809

@leebyron comments from original PR (edited, hopefully correctly):

> Fixes #3051
> 
> This change solves the problem of default values defined via SDL not
always resolving correctly through introspection by preserving the
original GraphQL literal in the schema definition. This changes argument
and input field definitions `defaultValue` field from just the "value"
to a new `GraphQLDefaultValueUsage` type which contains either (EDIT:
but not both!) "value" and "literal" fields.
> 
> Here is the flow for how a default value defined in an SDL would be
converted into a functional schema and back to an SDL:
> 
> **Before this change:**
> 
> ```
> (SDL) --parse-> (AST) --coerceInputLiteral--> (defaultValue config)
--valueToAST--> (AST) --print --> (SDL)
> ```
> 
> `coerceInputLiteral` performs coercion which is a one-way function,
and `valueToAST` is unsafe and set to be deprecated in #3049.
> 
> **After this change:**
> 
> ```
> (SDL) --parse-> (defaultValue literal config) --print --> (SDL)
> ```

Co-authored-by: Lee Byron <lee.byron@robinhood.com>

twof committed:
Remove CCN (#3999)

We're winding down work on CCN in favor of True Schema Nullability.

More context: https://github.com/graphql/nullability-wg/issues/37

andimarek committed:
Improve directive tests (#3775)

Improving directive tests by adding a new example for directives used on
other directive arguments:

```graphql
          directive @myDirective(arg:String) on ARGUMENT_DEFINITION 
          directive @myDirective2(arg:String @myDirective) on FIELD 
```

As far as I could see that was not clearly documented through tests
before.

yaacovCR committed:
fix: lint errors from PR prior to latest checks (#4204)

yaacovCR committed:
Preserve sources of variable values (#3811)

[#3077 rebased on
main](https://github.com/graphql/graphql-js/pull/3077).

Depends on #3810

@leebyron comments from original PR:

> By way of introducing type `VariableValues`, allows
`getVariableValues` to return both the coerced values as well as the
original sources, which are then made available in `ExecutionContext`.
> 
> While variable sources are not used directly here, they're used
directly in #3065. This PR is pulled out as a pre-req to aid review

---------

Co-authored-by: Lee Byron <lee.byron@robinhood.com>

yaacovCR committed:
Add valueToLiteral() (#3812)

[#3065 rebased on
main](https://github.com/graphql/graphql-js/pull/3065).

Depends on #3811

@leebyron comments from original PR:

> **Provides the "Value to Literal" methods in this [data flow
chart](https://user-images.githubusercontent.com/50130/118379946-51ac5300-b593-11eb-839f-c483ecfbc875.png).**
> 
> * Adds `valueToLiteral()` which takes an external input value and
translates it to a literal, allowing for custom scalars to define this
behavior.
> 
> **This also adds important changes to Input Coercion, especially for
custom scalars:**
> 
> * The value provided to `parseLiteral` is now `ConstValueNode` and the
second `variables` argument has been removed. For all built-in scalars
this has no effect, but any custom scalars which use complex literals no
longer need to do variable reconciliation manually (in fact most do not
-- this has been an easy subtle bug to miss).
>   This behavior is possible with the addition of `replaceVariables`

Changes to the original:
1. Instead of changing the signature of `parseLiteral()`, a new method
`parseConstLiteral()` has been added with the simpler signature.
`parseLiteral()` has been marked for deprecation.
2. `replaceVariables()` has access to operation and fragment variables.

Co-authored-by: Lee Byron <lee.byron@robinhood.com>

yaacovCR committed:
improve upgrade path for new custom scalar `parseConstLiteral()` method (#4209)

`parseLiteral()` has been marked as deprecated, prompting but not
forcing our users to convert to `parseConstLiteral()`.

With this additional change:

- in v17, if not supplied, the new `parseConstLiteral()` method will be
left as undefined, and during execution, we will fall back
to`parseLiteral()`.
- in v18, we will remove `parseLiteral()` and set up a default function
for `parseConstLiteral()` when not supplied.

Prior to this change, users of custom scalars with custom
`parseLiteral()` functions who did not provide a custom
`parseConstLiteral()` function would just get the default
`parseConstLiteral()` function during execution, which might not work as
expected.

This scheme will work except for users of custom scalars who want to
embed experimental fragment variables in their custom scalars, which
will only work with the new `parseConstLiteral()` method.

Cito committed:
Test suggestionList more thoroughly (#3241)

Particularly, verify that it first sorts by lexicographic distance and then naturally.

yaacovCR commented on src/type/validate.ts:
```diff
@@ -480,7 +480,20 @@ function validateUnionMembers(
+        );
+        includedInvalidTypes.add(typeString);
+      }
+      continue;
```
this reminds me of https://github.com/graphql/graphql-js/pull/4181 => what do you think about changing the error message string to something that references both errors, at least if there are repetitions?

'Something like Union type can only include unique Object types, ....'

What do you think?

yaacovCR reviewed (commented):
Seems like a good idea, left some comments/questions!
